File: client/functions/lib/functions/on_request/extend_cloud_task_scheduler.dart
import 'dart:async';

import 'package:junto_functions/functions/on_request_method.dart';
import 'package:junto_functions/utils/scheduled_functions.dart';
import 'package:junto_models/cloud_functions/requests.dart';

/// This class reschedules cloud tasks that exceed the 30 day quota for cloud tasks scheduling.
class ExtendCloudTaskScheduler extends OnRequestMethod<ExtendCloudTaskSchedulerRequest> {
  ExtendCloudTaskScheduler()
      : super(
          'ExtendCloudTaskScheduler',
          (jsonMap) => ExtendCloudTaskSchedulerRequest.fromJson(jsonMap),
        );

  @override
  Future<String> action(ExtendCloudTaskSchedulerRequest request) async {
    await scheduledFunctions.enqueueCall(
      request.functionName,
      request.payload,
      request.scheduledTime,
    );

    return '';
  }
}

File: firebase/functions/src/utils/infra/extend_cloud_task_scheduler.ts
import { OnRequestMethod } from '../../on_request_method'
import { scheduledFunctions } from './scheduled_functions'
import { ExtendCloudTaskSchedulerRequest } from '../../types'

/**
 * Reschedules Cloud Tasks that exceed the 30-day scheduling quota.
 * Corresponds to Dart's ExtendCloudTaskScheduler.
 */
export class ExtendCloudTaskScheduler extends OnRequestMethod<ExtendCloudTaskSchedulerRequest> {
    constructor() {
        super('ExtendCloudTaskScheduler', (jsonMap) => jsonMap as ExtendCloudTaskSchedulerRequest)
    }

    async action(request: ExtendCloudTaskSchedulerRequest): Promise<string> {
        await scheduledFunctions.enqueueCall(
            request.functionName,
            request.payload,
            new Date(request.scheduledTime)
        )
        return ''
    }
}
